sdks/go/pkg/beam/runners/direct/gbk.go

```diff
@@ -100,6 +114,68 @@ func (n *CoGBK) FinishBundle(ctx context.Context) error {
 	return n.Out.FinishBundle(ctx)
 }
 
+func (n *CoGBK) mergeWindows() error {
+	sort.Slice(n.wins, func(i int, j int) bool {
+		return n.wins[i].MaxTimestamp() < n.wins[j].MaxTimestamp()
+	})
+	n.mergeMap = make(map[typex.Window]int)
+	mergedWins := []typex.Window{}
+	for i := 0; i < len(n.wins); {
+		intWin, ok := n.wins[i].(window.IntervalWindow)
+		if !ok {
+			return errors.Errorf("tried to merge non-interval window type")
+		}
+		mergeStart := intWin.Start
+		mergeEnd := intWin.End
+		j := i + 1
+		for j < len(n.wins) {
+			candidateWin := n.wins[j].(window.IntervalWindow)
+			if candidateWin.Start <= mergeEnd {
+				mergeEnd = candidateWin.End
+				j++
+			} else {
+				break
+			}
+		}
+		for k := i; k < j; k++ {
+			n.mergeMap[n.wins[k]] = len(mergedWins)
+		}
+		mergedWins = append(mergedWins, window.IntervalWindow{Start: mergeStart, End: mergeEnd})
+		i = j
+	}
+	n.wins = mergedWins
+	return nil
+}
+
+func (n *CoGBK) reprocessByWindow() error {
+	newGroups := make(map[string]*group)
+	for _, g := range n.m {
+		ws := []typex.Window{n.wins[n.mergeMap[g.key.Windows[0]]]}
+		var buf bytes.Buffer
```

Comment:
Consider a helper function or two to reduce the duplication between this and ProcessElement. 
Ideally we simply merge the two paths (so delaying work in process element to FinishBundle), but until then, better to clean things up a little via helper functions or helper types. 

In this case, for example, the main difference is one is using `m.g` and this one is using `newGroups`. A helper type for grouping `type grouper map[string]*group` could handle this iteration in a method, avoiding the duplication. It just needs the right window information passed in somehow, in both cases... But two or so helper methods would be good, one for the key encoding and one for inserting into the group.  Since the method receiver is a map technically, it should use a value rather than a pointer, as maps, even as underlying types are reference types.

Reply:
Added helper functions for the key encoding and getting the group from the maps.